Provision DSA devices and workqueues through initcontainer

ozhuraki opened the pull request:
This PR is adds support for provisioning DSA devices and workqueues with accel-config utility for intel-dsa-plugin 
through initcontainer.

Closes: #658

mythi commented:
let's wait with this a bit still.

mythi commented on build/docker/intel-dsa-initcontainer.Dockerfile:
```diff
@@ -0,0 +1,27 @@
+RUN apt-get update && apt-get install -y uuid libjson-c5 kmod udev
+
+COPY --from=builder /usr/lib64/libaccel-config.so.1 /lib/x86_64-linux-gnu/
+COPY --from=builder /usr/lib64/libaccel-config.so.1.0.0 /lib/x86_64-linux-gnu/
```
`ldconfig` should create `libaccel-config.so.1.0.0` as a symlink?

mythi commented on build/docker/intel-dsa-initcontainer.Dockerfile:
```diff
@@ -0,0 +1,27 @@
+
+ARG ACCEL_CONFIG_VERSION=v2.8
+
+RUN wget -O- https://github.com/intel/idxd-config/archive/accel-config-$ACCEL_CONFIG_VERSION.tar.gz | tar -zx
```
add checksum

mythi commented on pkg/controllers/dsa/controller.go:
```diff
@@ -193,6 +221,12 @@ func (c *controller) NewDaemonSet(rawObj client.Object) *apps.DaemonSet {
 	}
+
+	// add the optional InitImage
+	if devicePlugin.Spec.InitImage != "" {
```
`InitImage` validation should be added in `dsadeviceplugin_webhook.go`

ozhuraki commented:
@mythi 

Added a license to DSA initcontainer Dockerfile. Will fix everything as you suggested, thanks for the input.

ozhuraki commented:
@mythi 

- Switched initcontainer to debian:buster-slim
- Added checksum for accel-config archive
- Fixed libaccel-config.so symlink issue
- Added InitImage validation to dsadeviceplugin_webhook.go

mythi commented on pkg/apis/deviceplugin/v1/dsadeviceplugin_types.go:
```diff
@@ -28,6 +28,9 @@ type DsaDevicePluginSpec struct {
 	Image string `json:"image,omitempty"`
 
+	// InitImage is an initcontainer image to provision DSA devices with accel-config utility.
+	InitImage string `json:"initImage,omitempty"`
```
We should re-think the name.  If we wanted something similar for, e.g., `GpuDevicePlugin`, this entry is already taken. `ProvisioningImage`?

rojkov replied:
IIUC the purpose of `InitImage` is to configure the host (e.g. install tools and drivers) and `ProvisioningImage` configures the devices on the host without touching the host's file system.

I'm thinking if this difference justifies having two separate images for a pod's init containers.

ozhuraki replied:
@mythi @rojkov

> We should re-think the name. If we wanted something similar for, e.g., GpuDevicePlugin, this entry is already taken. ProvisioningImage?

Regardless of the name it would translate to initcontainers in the deployment?

Implemented either way, it's better to document it clearly. Currently we specify it as "InitImage is an initcontainer image to provision DSA devices with accel-config utility".

As a first step I would simply update it to something like "InitImage / ProvisioningImage is an initcontainer image to configure and enable DSA devices with accel-config utility", enhancing the details if neccessary.

> IIUC the purpose of InitImage is to configure the host (e.g. install tools and drivers) and ProvisioningImage configures the devices on the host without touching the host's file system.

The split in between configuring the host / provisioning SW/HW might be fuzzy (also to understand).

> I'm thinking if this difference justifies having two separate images for a pod's init containers.

My 2 cents: I am fine with ProvisioningImage. I would say there's no right/wrong approach here, I would prefer  a simpler and less confusing solution.

I will update to ProvisioningImage and try to make description less ambiguous.

mythi replied:
> The split in between configuring the host / provisioning SW/HW might be fuzzy (also to understand).

Can we do both easily with just one image? I'm also fine with one `InitImage` entry per CRD that is tasked to do both if what it needs to do can be customized somehow (e.g. not install some tools but provision only vs do both).

>> I'm thinking if this difference justifies having two separate images for a pod's init containers.
>
> My 2 cents: I am fine with ProvisioningImage. I would say there's no right/wrong approach here,
> I would prefer a simpler and less confusing solution.

+1, the user experience is what matters.

mythi replied:
`InitImage` is OK

ozhuraki commented:
@mythi @rojkov
- InitImage -> ProvisioningImage
- Updated the description to "ProvisioningImage is an initcontainer image to configure and enable DSA devices and workqueues with accel-config utility"

ozhuraki commented:
@mythi 

Oops, missed your comments. I will update.
> Can we do both easily with just one image? I'm also fine with one InitImage entry per CRD that is tasked to do both if what it needs to do can be customized somehow (e.g. not install some tools but provision only vs do both).

+1
I think having multiple images could be justified in cases where it's not possible due to non-technical reasons, etc (or where it would simplify the task). In this specific case, there aren't such yet.

ozhuraki commented:
@mythi
- ProvisioningImage -> InitImage

mythi commented on build/docker/intel-dsa-initcontainer.Dockerfile:
```diff
@@ -0,0 +1,50 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+FROM debian:buster-slim AS builder
```
similar to #696 I'd suggest `sid-slim`

mythi commented on build/docker/intel-dsa-initcontainer.Dockerfile:
```diff
@@ -0,0 +1,50 @@
+
+FROM debian:buster-slim
+
+RUN apt-get update && apt-get install -y uuid libjson-c3 kmod udev
```
`libjson-c4` from `sid`

mythi commented on build/docker/intel-dsa-initcontainer.Dockerfile:
```diff
@@ -0,0 +1,50 @@
+COPY --from=builder /usr/bin/accel-config /usr/bin/
+COPY --from=builder /test /test
+
+ENTRYPOINT /bin/mount -o remount,rw /sys && /usr/bin/accel-config load-config -c /test/configs/2g2q_user_1.conf && /usr/bin/accel-config enable-device dsa0 && /usr/bin/accel-config enable-wq dsa0/wq0.0 && /usr/bin/accel-config list
```
let's try to revisit this `ENTRYPOINT` and by default enable *all* available DSA devices

bart0sh replied:
Can we use volume mounts instead of remounting /sys ?

ozhuraki replied:
@bart0sh 
> Can we use volume mounts instead of remounting /sys ?

If I am not missing anything, we do:

https://github.com/intel/intel-device-plugins-for-kubernetes/blob/9545146ae96d01100085d7ad74d6553858492f57/deployments/dsa_plugin/overlays/dsa_initcontainer/dsa_initcontainer.yaml#L1-L20

I am not quite sure where is the problem, need to dig into this.

mythi reviewed (changes requested):
a few additional comments

ozhuraki commented:
Changes from the previous version:

- Switched DSA initcontainer image to debian:sid-slim
- Updated DSA initcontainer image to use libjson-c5
- Upgrated accel-config to 3.4.1

TODO/WIP

- Check if remount of /sys could be dropped
- Enable all idle DSA devices

mythi commented on build/docker/intel-dsa-initcontainer.Dockerfile:
```diff
@@ -0,0 +1,50 @@
+            pkgconf asciidoc xmlto uuid-dev libjson-c-dev libkmod-dev \
+            libudev-dev libkeyutils-dev curl
+
+ENV ACCEL_CONFIG_VERSION 3.4.1
+ENV ACCEL_CONFIG_DOWNLOAD_URL https://github.com/intel/idxd-config/archive/accel-config-v$ACCEL_CONFIG_VERSION.tar.gz
+ENV ACCEL_CONFIG_SHA256 8e924ce4d6d75819e9a25e11624f7c603e86c37297bc1c7af53aed48b7f9e0fa
```
these should be `ARG`s

ozhuraki commented:
Changes:
- ENV -> ARG
- accel-config -> 3.4.2
- debian:sid-slim -> debian:unstable
- Dropped remount
- Dropped all volumes, using just /sys/devices
- Added enabling of all idle DSA devices/workqueues (1 engine / 1 group / 1 wq (user/dedicated) )

mythi commented on demo/dsa-init.sh:
```diff
@@ -0,0 +1,33 @@
+
+    dev="dsa${i}"
+
+    sed "s/X/${i}/g" < dsa.conf > $dev.conf
```
this fails when we have `readOnlyRootFilesystem: true`

mythi commented on build/docker/intel-dsa-initcontainer.Dockerfile:
```diff
@@ -0,0 +1,52 @@
+    ./configure CFLAGS='-g -O2' --prefix=/usr --sysconfdir=/etc --libdir=/usr/lib64 --disable-test --disable-docs && \
+    make && \
+    make install && \
+    mkdir /test && \
+    cp -R test/configs /test/
```
not needed

ozhuraki commented:
Changes:
- Dropped "readonlyrootfs" from DSA initcontainer deployment and operator
- Added /sys/devices volume into DSA initcontainer deployment

ozhuraki commented:
Changes:
- Dropped /test from initcontainer

ozhuraki commented:
Changes:
- Added "WORKDIR /dsa-init" to initcontainer

ozhuraki commented:
Changes:
- Added /sys/devices volume into DSA operator

bart0sh commented:
@ozhuraki 3 more things would be nice to have:
- documentation (overall description, deployment instructions, etc)
- list of known issues
- github issue(s) for the known issue(s) included into the [next project milestone](https://github.com/intel/intel-device-plugins-for-kubernetes/projects/23)

mythi commented:
> - list of known issues
> - github issue(s) for the known issue(s) included into the next project milestone

I can cover these in the release notes but a short description and sample

```
$ kubectl apply -k deployments/dsa_plugin/overlays/dsa_initcontainer/ 
```

would indeed be good in the README

mythi reviewed (approved):
this works and is good IMO. short README update as @bart0sh suggested would be good still.

ozhuraki commented:
@mythi @bart0sh 

Added a documentation, please take a look.

mythi commented:
LGTM. Let's merge

rojkov reviewed (approved):
LGTM
